### ISSUE TITLE
feat: support more validation libraries

### ISSUE THREAD

**Issue by juliusmarminge** [t3-oss/t3-env#6]
Instead of 
```ts
createEnv({
  server: {
    KEY: z.string(),
  }
});
```

it should be 
```ts
createEnv({
  server: z.object({
    KEY: z.string(),
  })
});
```

and it should not be tied to zod, meaning something like [scale-ts](https://github.com/paritytech/scale-ts) could work for example:
import * as $ from 'scale-codec';

```ts
createEnv({
  server: $.object(
    $.field('text', $.str)
  ),
});
```

this would require some more considerations for typings as we'd need to pull out the shape of each to get access to the raw keys to compare against the client prefix

Should probably implement #169 first in some half-generic way 🤔

**Comment by chungweileong94** [comment]
Probably need an adapter for each validator, for couple of reasons:
- To tell the core how to extract the typing, mainly for validating client prefix, and type infer in general
- To tell the core how to parse the env values. As we call `.safeParse` in zod, but we need to do it in different way for other validator.

**Comment by juliusmarminge** [comment]
We support a bunch of validators in tRPC:

https://github.com/trpc/trpc/blob/main/packages/server/src/core/parser.ts

https://github.com/trpc/trpc/blob/main/packages/server/src/core/internals/getParseFn.ts

**Comment by taylorallen0913** [comment]
Can I look into integrating Yup as an option alongside Zod?

**Comment by juliusmarminge** [comment]
sure!

**Comment by chungweileong94** [comment]
@aalakoski Most probably no, the ideal solution is to make t3-env understand how to extract things from different validators without having bundle with any of the validator packages.

**Comment by SeanCassiere** [comment]
Tanner's pattern of [validating the router search params](https://github.com/TanStack/router/blob/beta/examples/react/kitchen-sink-multi-file/src/routes/dashboard/users/index.tsx#L30-L37), could be something to consider here.

The process in t3-env would look like this.
1. Make the `client` and `server` keys in the init of `createEnv ` accept validation functions, that receive the values of 
 `process.env` (or `import.meta`, ... etc.) as its argument.
2. Then call the user-defined `client` and `server` validation functions to the ensure the values are adhere to the schema.

Code-example with zod:
```ts
import { createEnv } from "@t3-oss/env-core";
import { z } from "zod";

const env = createEnv({
  server: z.object({ FOO: z.string().min(1) }).parse,
  client: z.object({ BAR: z.string().min(1) }).parse
});
```


Code example with a custom validator
```ts
import { createEnv } from "@t3-oss/env-core";

// custom validator which could be swapped for Yup, Joi, etc....
type ServerVars = { FOO: string }
function serverValidator(values: any): ServerVars {
  // validation magic
  return { FOO: "FOOOOOOOOO" }
};

type ClientVars = { BAR: string }
function clientValidator(values: any): ClientVars {
  // validation magic
  return { BAR: "BOOOOOOOOO" }
};

const env = createEnv({
  server: serverValidator,
  client: clientValidator
});
```

This pattern completely opens it up to the user to bring their own validation libraries and logic into the picture, with `t3-env` making use of the type-safety returned by the defined validators as well as the libs' client-server separation stuff.

**Comment by juliusmarminge** [comment]
> Will it also export zod / other validators from within t3-env or would you still need to add zod as a dependency? If not, could it be done?
> 
> Something like:
> 
> ```
> import { createEnv, z } from "@t3-oss/env-core/zod";
> 
> const env = createEnv({
>   server: {
>     FOO: z.string(),
>   }
> })
> ```

No - we won't bundle the validator

**Comment by Bekacru** [comment]
why does it need validation library tho isn't is simple enough task to do plainly? what am I missing?

**Comment by mattddean** [comment]
Did a very naiive [valibot](https://github.com/fabian-hiller/valibot) replacement.
Obviously it doesn't accomplish the goal here of not shipping a validation library, but it might be useful to someone.

**Comment by KaptinLin** [comment]
any plan support more validation libraies? i want use [valibot](https://github.com/fabian-hiller/valibot) instead of zod

**Comment by estubmo** [comment]
> Did a very naiive [valibot](https://github.com/fabian-hiller/valibot) replacement. Obviously it doesn't accomplish the goal here of not shipping a validation library, but it might be useful to someone. https://github.com/mattddean/t3-env-valibot/tree/feature/valibot

This seems to be no longer working, as parse is no longer a function of schemas.

**Comment by stunaz** [comment]
why not using the patterns as in https://github.com/react-hook-form/resolvers/tree/master ?

**Comment by chungweileong94** [comment]
> why not using the patterns as in https://github.com/react-hook-form/resolvers/tree/master ?

Yeah, maybe we have to, as Valibot provides a separate `parse` method instead of from the scheme. Of course I also aware that they do have method from the scheme called `_parse`, but it's for internal use only, and I heard that it only provides some raw behavior.

**Comment by AlexisWalravens** [comment]
I just stumbled upon this issue, and maybe this could be useful https://typeschema.com ?
[next-safe-action](https://next-safe-action.dev/docs/introduction) just implemented it in it's last version, allowing people to use almost every schema validator.

**Comment by Talent30** [comment]
> I just stumbled upon this issue, and maybe this could be useful https://typeschema.com ? [next-safe-action](https://next-safe-action.dev/docs/introduction) just implemented it in it's last version, allowing people to use almost every schema validator.

but it only supports async parse

**Comment by decs** [comment]
would you guys be open to making `createEnv` an async function? there are some considerations regarding top-level awaits, but maye we can figure something out

**Comment by juliusmarminge** [comment]
I think we'd make a `createEnvAsync` in that case

**Comment by chungweileong94** [comment]
I very interested in getting my hand dirty on this one, but the only problem is that how will the present is going to fit in, as now all the presents are defined with zod schema.

**Comment by juliusmarminge** [comment]
> I very interested in getting my hand dirty on this one, but the only problem is that how will the present is going to fit in, as now all the presents are defined with zod schema.

We can have entrypoint for each `@t3-oss/env-core/preset-zod` for example

**Comment by fabian-hiller** [comment]
A few schema library authors are currently working on a common interface. This may allow this library to support any schema library that supports this interface in the future. This will simplify the integration of schema libraries and prevent vendor lock-in, which could lead to more innovation in the long run.

**Comment by chungweileong94** [comment]
> A few schema library authors are currently working on a common interface.

So this is the common interface https://github.com/standard-schema/standard-schema, which is pretty cool. 
However, zod hasn’t adapted to it just yet, so I think we either wait or implement an adapter with `standard-schema`.

**Comment by Gakai** [comment]
As of v3.24 (Dec 10), zod also implements the standard-schema

**Comment by juliusmarminge** [comment]
Standard Schema will be supported in 0.12
